### ISSUE TITLE
Feature Request: Haptic Feedback (force feedback, rumble)

### ISSUE THREAD

**Issue by Areopagitics** [efroemling/ballistica#174]
Hi Eric, 

Your game would be great with force feedback. I lfound that sdl2, the one you are using, has force feedback.  

https://wiki.libsdl.org/CategoryForceFeedback

Let me know if I could help in some way. I have a bit of experience with python, not too much with C++ but wouldn't mind getting my hands a bit dirty ... would love to help if it's not too complicated.

**Comment by Areopagitics** [comment]
It looks like you would just need to call [SDL_HapticRumbleSupported ](https://wiki.libsdl.org/SDL_HapticRumbleSupported)on the controller to check if it has rumble (when it connects), and then [SDL_HapticRumblePlay](https://wiki.libsdl.org/SDL_HapticRumblePlay) triggered by events (for example when receiving punches, bombs, or snowballs).

**Comment by efroemling** [comment]
I would be all for adding something like this, though its low on my priority list at the moment. If anyone wants to help out, the first step would be adding some sort of haptic events to the InputDevice class (perhaps virtual functions that by default do nothing). Then the Joystick class could override those to talk to SDL's functionality as you mention above. One could test it by simply having joysticks rumble when buttons are pressed on them/etc. but after that we would want to wire it up to be controlled by game events coming from the python layer.

**Comment by ritiek** [comment]
Interesting. Let me see if I can help you get this to work!

**Comment by Areopagitics** [comment]
@ritiek let me know if I can do any testing. I have both USB and Bluetooth controllers with vibration.

**Comment by ritiek** [comment]
@Areopagitics Thanks! I only have a Dualshock 4 controller which can be used both as wired and wireless. It'd be nice to see what happens with other controllers too.

I made a very experimental change on my fork in [ritiek/ballistica@joystick-rumble-support](https://github.com/ritiek/ballistica/tree/joystick-rumble-support). The controller should vibrate a little whenever you press the punch button in-game.

**Comment by Areopagitics** [comment]
sweet! I'll do a pull of your fork and try building on my Windows for the first time soon.  I finally found the [instructions ](https://github.com/efroemling/ballistica/wiki/Getting-Started#running); they're great.

**Comment by Areopagitics** [comment]
@ritiek I'm guessing I would need to do a full clean build from source. I tried your repo using the prefab command for windows; I was able to build it with some errors and then run, but no luck getting the rumble to work. I tested the rumble on USB controller configuration settings though, and it's working fine. 

Any suggestions on how to build on a WSL 2 setup on windows?

**Comment by SyedAhkam** [comment]
Hey, whats the status on this? I'd love to help out since this feature sounds quite fun to implement. 

I do have a DS4 lying around as well.

@ritiek @Areopagitics

**Comment by ritiek** [comment]
Hi @SyedAhkam.

I ported my above mentioned changes from 2020 to the latest version of the game as of few months ago. This should be a good starting point: [ritiek/ballistica@game-controller-rumbling](https://github.com/ritiek/ballistica/tree/game-controller-rumbling) (should rumble when you press punch button on the controller)

I remember it working fine on my Linux machine. Don't have Windows around to test with, so not sure if or how to get these changes working on Windows.

Ideally in the actual game, I feel it should rumble on say when our punch actually lands on the opponent (not just on pressing the punch button, which is what I did in my branch up), or when our character gets punched, or there's a bomb detonating somewhere. So someone will have to research where all in the codebase should we be adding these rumble calls. It'd also be nice to make sure this rumble behaviour works the same way during online play.

**Comment by brostosjoined** [comment]
ill be happy to test it out for you on windows when am free @ritiek

**Comment by MatiasMGz** [comment]
Still in the low priority list, right? Huh... i would like to test it on Windows.

**Comment by SyedAhkam** [comment]
@brostosjoined @MatiasMGz 

Hi, one of you guys can go ahead and build Ritiek's fork from source, Try it out on windows and report back here please.

**Comment by MatiasMGz** [comment]
Understood, sir!

**Comment by MatiasMGz** [comment]
I hope I can test this between the next week!

**Comment by brostosjoined** [comment]
@ritiek update the fork 

```
pcommandbatch got negative age; unexpected.Error: Server gave an error. Old build files may no longer be available on the server; make sure you are using a recent commit.
Note that build files will remain available indefinitely once downloaded, even if deleted by the server. So as long as your .cache/efrocache directory stays intact you should be able to repeat any builds you have run before.
```

**Comment by ritiek** [comment]
Apologies. I'm afraid I won't be able to update my fork to work with the
newest changes any time soon.

Feel free to refer to or clone the work done in my branch so far, whoever's
feeling up to it.

>

**Comment by MatiasMGz** [comment]
Oh...

**Comment by MatiasMGz** [comment]
:(

**Comment by MatiasMGz** [comment]
So... could we use SDL3? Anyway, i am still trying it just in case. :b

**Comment by efroemling** [comment]
Yes I think we should adopt SDL3 before adding haptics so we're using the latest and greatest stuff. SDL3 was just officially released a few days ago, so I want to give it a bit of time to stabilize and get out to various package managers and then we can upgrade to it. Also there'll be some backwards-incompatible changes as part of 1.8 (for asset downloading, etc) so maybe that would be a good time to also introduce haptics.

**Comment by sedbytes** [comment]
Why did we close this, what is the status on SDL3 migration?

**Comment by NewtonChutney** [comment]
There doesn't seem to have been an associated commit that closed this issue.. It was done manually..

Please do notify the community about any intentions from the project @Loup-Garou911XD

**Comment by Loup-Garou911XD** [comment]
> There doesn't seem to have been an associated commit that closed this issue.. It was done manually..
> 
> Please do notify the community about any intentions from the project @Loup-Garou911XD 

Still no work on this
I closed it as it was stale but since u guys are still active we should keep it open

**Comment by efroemling** [comment]
Yeah let's keep this open.

Quick update on the state of the world here:

SDL3 is out now (woohoo!) but it doesn't seem like it'll be widely available in linux distros like Ubuntu until their next major release, so I don't think it would be wise to simply switch over to it right now. Though if someone wanted to fork the existing sdl2 support code into sdl2 and sdl3 options I'd be happy to accept a PR. Then we could switch the Windows build to sdl3 and linux folks could use either (with sdl2 probably remaining the default for a while). If anyone is interested in tackling this, ping me and I can provide details on what it'd require.

I think it'd make sense to implement improvements such as haptics in the sdl3 backend (and folks can backport stuff to sdl2 when it makes sense).

The one other important thing we need for haptics is a protocol update - the server needs to be able to tell clients when they should do haptics stuff on what controllers, etc. The upcoming 1.8 release will be bumping the protocol for asset-packages and other backwards-incompatible stuff so this would be a convenient time to add some sort of haptics messaging too. So please remind me to add this before 1.8 final rolls out if I forget.

**Comment by DinoWattz** [comment]
Also please don't forget to make the controller rumble only when the spaz being controlled gets damaged or something, 'cause if it rumbles for every single bomb exploding in the screen it'd be a nightmare to play.. the camera shake could probably be improved as well..

**Comment by Loup-Garou911XD** [comment]
the game switched to sdl3, just waiting for this to be implemented 😌

**Comment by Loup-Garou911XD** [comment]
@efroemling maybe something u could do in a jiffy

**Comment by sedbytes** [comment]
I wrote an implementation in a jiffy with some AI assistance. The PR is out @ #943

@/all please test it out if possible.

@efroemling please review :)
